hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java

```diff
@@ -0,0 +1,341 @@
+                continue;
+            }
+
+            IndexIterationPointer pointer = pointers[i];
+
+            // For hash lookups, pointer begin and end points will be the same
+            assert pointer.getFrom() == pointer.getTo();
+
+            List<QueryableEntry> filteredEntries = getOwnedEntries(index.getRecords(pointer.getFrom()), partitionIdSet);
+            entries.addAll(filteredEntries);
+            newPointers[i] = IndexIterationPointer.createFinishedIterator();
+
+            if (entries.size() >= sizeHint) {
+                sizeHintReached = true;
+            }
+        }
+
+        return new ResultInternal(entries, newPointers);
+    }
+
+    @Override
+    public Object getResponse() {
+        return response;
+    }
+
+    @Override
+    protected void readInternal(ObjectDataInput in) throws IOException {
+        super.readInternal(in);
+        indexName = in.readString();
+        partitionIdSet = in.readObject();
+        pointers = in.readObject();
+        sizeHint = in.readInt();
+    }
+
+    @Override
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
+        super.writeInternal(out);
+        out.writeString(indexName);
+        out.writeObject(partitionIdSet);
+        out.writeObject(pointers);
+        out.writeInt(sizeHint);
+    }
+
+    @Override
+    public int getClassId() {
+        return MAP_FETCH_INDEX;
+    }
+
+    private boolean allIndexed(InternalIndex index, PartitionIdSet partitionIdSet) {
+        return partitionIdSet.stream().allMatch(index::hasPartitionIndexed);
+    }
+
+    protected int getMigrationTimestamp() {
+        return mapService.getMigrationStamp();
+    }
+
+    /**
+     *  Return only the {@code entries} from the partitions declared in {@code partitionIdSet}
+     *
+     * @param entries
+     * @param partitionIdSet
+     * @return list of entries belonging the given partitions
+     */
+    @Nonnull
+    private List<QueryableEntry> getOwnedEntries(
+            @Nonnull Collection<QueryableEntry> entries,
+            @Nonnull PartitionIdSet partitionIdSet
+    ) {
+        return entries.stream()
+                .filter(entry -> {
+                    int partitionId = partitionService.getPartitionId(entry.getKeyData());
+                    return partitionIdSet.contains(partitionId);
+                })
+                .collect(Collectors.toList());
+    }
+
+    public static class MapFetchIndexOperationResult {
+        private final Set<Integer> migratedPartitionIds;
+        private final ResultInternal inner;
+
+        public MapFetchIndexOperationResult(
+                List<QueryableEntry> result,
+                Set<Integer> migratedPartitionIds,
+                IndexIterationPointer[] pointers
+        ) {
+            this.migratedPartitionIds = migratedPartitionIds;
+            this.inner = new ResultInternal(result, pointers);
+        }
+
+        public List<QueryableEntry> getResult() {
```

Comment:
Why `QueryableEntry` doesn't indicate generic parameters?

Reply:
Also, I haven't see any `MissingPartitionException` usage :)

Reply:
Unfortunately, we don't know the inner types by just looking at index structure. So we lose info on the generic type instantiations.

For your second comment: yes, you are right. I will add that particular exception.